lxd/firewall/firewall_load.go

```diff
@@ -10,26 +11,35 @@ func New() Firewall {
 	nftables := drivers.Nftables{}
 	xtables := drivers.Xtables{}
 
-	// If nftables is compatible and already in use, then we prefer to use the nftables driver irrespective of
-	// whether xtables is in use or not.
-	nftablesCompat, nftablesInUse := nftables.Compat()
-	if nftablesCompat && nftablesInUse {
+	nftablesInUse, nftablesCompatErr := nftables.Compat()
+	if nftablesCompatErr != nil {
+		logger.Debugf(`Firewall detected "nftables" incompatibility: %v`, nftablesCompatErr)
+	} else if nftablesInUse {
+		// If nftables is compatible and already in use, then we prefer to use the nftables driver
+		// irrespective of whether xtables is in use or not.
 		return nftables
-	} else if !nftablesCompat {
-		// Note: If nftables isn't compatible, we fallback to xtables without considering whether xtables
-		// is itself compatible. This continues the existing behaviour of allowing LXD to start with
-		// potentially an incomplete firewall backend, so that only networks and instances using those
-		// features will fail to start later.
-		return xtables
 	}
 
-	// If xtables is compatible and already in use, then we prefer to stick with the xtables driver rather than
-	// mix the use of firewall drivers on the system.
-	xtablesCompat, xtablesInUse := xtables.Compat()
-	if xtablesCompat && xtablesInUse {
+	xtablesInUse, xtablesCompatErr := xtables.Compat()
+	if xtablesCompatErr != nil {
+		logger.Debugf(`Firewall detected "xtables" incompatibility: %v`, xtablesCompatErr)
+	} else if xtablesInUse {
+		// If xtables is compatible and already in use, then we prefer to stick with the xtables driver
+		// rather than mix the use of firewall drivers on the system.
 		return xtables
 	}
 
-	// Otherwise prefer nftables as default.
-	return nftables
+	// If nftables is compatible, but not in use, and xtables is not compatible or not in use, use nftables.
+	if nftablesCompatErr == nil {
+		return nftables
+	}
+
+	// If neither nftables nor xtables are compatible, we fallback to xtables.
+	// This continues the existing behaviour of allowing LXD to start with potentially an incomplete firewall
+	// backend, so that only networks and instances using those features may fail to function properly.
+	// The most common scenario for this is when xtables is using nft shim commands but the nft command itself
+	// is not installed. In this case LXD will use the xtables shim commands but with the potential of problems
+	// due to differences between the original xtables commands and the shim commands provided by nft.
+	logger.Warnf(`Firewall failed to detect any compatible driver, falling back to "xtables" (but some features may not work as expected due to: %v)`, xtablesCompatErr)
```

Comment:
Did I miss something or will we be displaying this warning on systems where:
 - nftables isn't available
 - xtables is functional but not in use

Reply:
I believe the warning should only be displayed with both nftablesCompatErr and xtablesCompatErr are non-nil

Reply:
@stgraber yep you're right, good spot, it was like that original, but I removed it last at minute for some reason.

Reply:
@stgraber fixed.